### PR TITLE
PLT-1147: Relax float-in criteria

### PR DESCRIPTION
**Pull request by zliu41**
See added Note for details.

### CODE REVIEW

File: plutus-tx-plugin/test/Budget/Spec.hs [hunk 1]
```diff
@@ -155,3 +160,25 @@ compiledShow :: CompiledCode Integer
 compiledShow = $$(compile [||
       let x = -1234567890
        in showExample x ||])
+
+-- | In this example, the float-in pass cannot reduce the cost unless it allows
+-- unconditionally floating into type abstractions. Both branches are
+-- turned into type abstractions (because the `a + a` branch is not a value).
```
Comment (michaelpj) [inline plutus-tx-plugin/test/Budget/Spec.hs:166]: That doesn't seem right? The branches aren't in a multi-occurence context so floating in should have been fine already? Same applies below. Did you observe a difference when turning it on?

File: plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        bad_name
+        (all
+          a
+          (type)
+          (fun [ (lam a (type) (fun a (fun a a))) a ] (fun a (fun a a)))
+        )
+      )
+      (abs a (type) (lam v [ (lam a (type) (fun a (fun a a))) a ] v))
+    )
+    (termbind
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden:26]: Perhaps we're using `compileToReadable` so the optimization doesn't fire yet? Unclear.

File: plutus-tx-plugin/test/Budget/Spec.hs (hunk 1 above)
Comment (zliu41) [reply plutus-tx-plugin/test/Budget/Spec.hs:166]: When floating into the argument of an application, it is considered multi-occ context, unless the function is either `\x -> ...` where `x` occurs once, or a builtin. If the function is a `Let` or `App` etc. then it is considered multi-occ context. Do you have a better approach?

File: plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden (hunk 2 above)
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden:26]: Exactly, float-in is after `compiledToReadable`

File: plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs [hunk 3]
```diff
@@ -386,19 +410,11 @@ floatInBinding ver letAnn = \b ->
...
         LamAbs (a, usBody) n ty lamAbsBody
             | Set.disjoint declaredUniqs (typeUniqs ty) ->
-                -- We float into lambdas only if `_ctxtInManyOccRhs = False`.
-                -- See Note [Float-in] #4
-                ifM
-                    (asks _ctxtInManyOccRhs)
-                    giveup
-                    (LamAbs (a, usBind <> usBody) n ty <$> go b lamAbsBody)
+                -- See Note [Float-in] #5
+                LamAbs (a, usBind <> usBody) n ty <$> go b lamAbsBody
```
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs:414]: > optimization never makes things worse

I don't think we can claim that. We don't have proof that float-in can't affect other passes. In fact it likely can: float-in can make RHSs bigger, which makes it harder to inline things (especially once we have a more sophisticated inliner).

File: plutus-tx-plugin/test/Budget/Spec.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-tx-plugin/test/Budget/Spec.hs:166]: Aha, so the multi-occ comes from the fact that the case expression is compiled to a function application, of course. Brain fart.

File: plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden:26]: Are you sure it is in fact floated in? :innocent:

File: plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs:414]: Why not? You're saying that float in can make the effect of later passes worse. What I'm saying is compatible with that: `conservative` should mean that optimization never makes things worse than _the original program_, which is a think that can definitely happen with non-conservative optimization. And we can get this property piecewise: if all of our passes are conservative, then the whole pipeline is conservative.

File: plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden (hunk 2 above)
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-tx-plugin/test/Budget/ifThenElse1.plc.golden:26]: Yep, I checked.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs:414]: I see, that makes sense.

[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs]: Add the flags.

File: plutus-core/plutus-ir/test/TransformSpec.hs [hunk 4]
```diff
@@ -133,7 +133,7 @@ letFloatIn =
             ]
   where
     goldenFloatTC pir = rethrow . asIfThrown @(PIR.Error PLC.DefaultUni PLC.DefaultFun ()) $ do
-        let pirFloated = runQuote $ LetFloatIn.floatTerm def pir
+        let pirFloated = runQuote $ LetFloatIn.floatTerm def True pir
```
Comment (michaelpj) [inline plutus-core/plutus-ir/test/TransformSpec.hs:136]: Okay, now I want _one_ test showing that non-relaxed float-in is in fact non-relaxed :)

File: plutus-tx-plugin/src/PlutusTx/Options.hs [hunk 5]
```diff
@@ -64,19 +66,25 @@ makeLenses ''PluginOptions
...
+    , poFun          :: Maybe OptionValue -> Validation ParseError (a -> a)
     -- ^ Consumes an optional value, and either updates the field or reports an error.
-    , poLens        :: Lens' PluginOptions a
+    , poLens         :: Lens' PluginOptions a
     -- ^ Lens focusing on the field. This is for modifying the field, as well as
     -- getting the field value from `defaultPluginOptions` for pretty printing.
-    , poDescription :: Text
+    , poDescription  :: Text
     -- ^ A description of the option.
+    , poImplications :: [Implication a]
```
Comment (michaelpj) [inline plutus-tx-plugin/src/PlutusTx/Options.hs:85]: cute

**Review by michaelpj — APPROVED** [review]

File: plutus-core/plutus-ir/test/TransformSpec.hs (hunk 4 above)
Comment (zliu41) [reply plutus-core/plutus-ir/test/TransformSpec.hs:154]: Right! Added a few test cases

**Pull request merged**